File: app/src/main/java/ca/appvelopers/mcgillmobile/ui/courses/CoursesActivity.java
/*
 * Copyright 2014-2017 Julien Guerinet
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package ca.appvelopers.mcgillmobile.ui.courses;

import android.content.DialogInterface;
import android.content.Intent;
import android.os.Bundle;
import android.support.v4.content.LocalBroadcastManager;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.LinearLayout;
import android.widget.TextView;

import com.guerinet.utils.Utils;
import com.guerinet.utils.dialog.DialogUtils;

import java.util.List;

import javax.inject.Inject;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;
import ca.appvelopers.mcgillmobile.App;
import ca.appvelopers.mcgillmobile.R;
import ca.appvelopers.mcgillmobile.RegistrationError;
import ca.appvelopers.mcgillmobile.model.Course;
import ca.appvelopers.mcgillmobile.model.Term;
import ca.appvelopers.mcgillmobile.model.Transcript;
import ca.appvelopers.mcgillmobile.model.exception.MinervaException;
import ca.appvelopers.mcgillmobile.ui.DrawerActivity;
import ca.appvelopers.mcgillmobile.ui.dialog.DialogHelper;
import ca.appvelopers.mcgillmobile.ui.dialog.list.TermDialogHelper;
import ca.appvelopers.mcgillmobile.util.Constants;
import ca.appvelopers.mcgillmobile.util.dbflow.databases.CoursesDB;
import ca.appvelopers.mcgillmobile.util.manager.HomepageManager;
import ca.appvelopers.mcgillmobile.util.manager.McGillManager;
import ca.appvelopers.mcgillmobile.util.manager.TranscriptManager;
import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;
import timber.log.Timber;

/**
 * Shows the user all of the courses the user has taken or is currently registered in
 * @author Julien Guerinet
 * @author Joshua David Alfaro
 * @since 1.0.0
 */
public class CoursesActivity extends DrawerActivity {
    /**
     * {@link Course}s list
     */
    @BindView(android.R.id.list)
    RecyclerView list;
    /**
     * Button to unregister from a course
     */
    @BindView(R.id.course_register)
    Button unregisterButton;
    /**
     * Empty list view
     */
    @BindView(R.id.courses_empty)
    TextView emptyView;
    /**
     * {@link TranscriptManager} instance
     */
    @Inject
    TranscriptManager transcriptManager;
    /**
     * Adapter for the {@link Course}s list
     */
    private CoursesAdapter adapter;
    /**
     * Current {@link Term} shown
     */
    private Term term;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_wishlist);
        ButterKnife.bind(this);
        App.component(this).inject(this);
        analytics.sendScreen("View Courses");

        term = App.getDefaultTerm();

        list.setLayoutManager(new LinearLayoutManager(this));

        //Remove this button
        findViewById(R.id.course_wishlist).setVisibility(View.GONE);
    }

    @Override
    protected void onResume() {
        super.onResume();
        update();
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.refresh, menu);
        getMenuInflater().inflate(R.menu.change_semester, menu);
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case R.id.action_change_semester:
                DialogUtils.list(this, R.string.title_change_semester,
                        new TermDialogHelper(this, term, false) {
                            @Override
                            public void onTermSelected(Term term) {
                                //Set the default term
                                App.setDefaultTerm(term);

                                //Set the instance term
                                CoursesActivity.this.term = term;

                                update();
                                refresh();
                            }
                        });
                return true;
            case R.id.action_refresh:
                refresh();
                return true;
            default:
                return super.onOptionsItemSelected(item);
        }
    }

    /**
     * Updates all of the info in the view
     */
    private void update() {
        //Set the title
        setTitle(term.getString(this));

        //User can unregister if the current term is in the list of terms to register for
        boolean canUnregister = App.getRegisterTerms().contains(term);

        //Change the text and the visibility if we are in the list of currently registered courses
        if (canUnregister) {
            unregisterButton.setVisibility(View.VISIBLE);
            unregisterButton.setText(R.string.courses_unregister);
            LinearLayout.LayoutParams params = new LinearLayout.LayoutParams(
                    ViewGroup.LayoutParams.MATCH_PARENT,
                    ViewGroup.LayoutParams.WRAP_CONTENT);
            unregisterButton.setLayoutParams(params);
        } else {
            unregisterButton.setVisibility(View.GONE);
        }

        //Set up the list
        adapter = new CoursesAdapter(emptyView, term, canUnregister);
        list.setAdapter(adapter);
    }

    /**
     * Refreshes the list of courses for the given term and the user's transcript
     */
    private void refresh() {
        if (!canRefresh()) {
            return;
        }

        //Download the courses for this term
        mcGillService.schedule(term).enqueue(new Callback<List<Course>>() {
            @Override
            public void onResponse(Call<List<Course>> call, Response<List<Course>> response) {
                // Set the courses
                CoursesDB.setCourses(term, response.body());

                //Download the transcript (if ever the user has new semesters on their transcript)
                mcGillService.transcript().enqueue(new Callback<Transcript>() {
                    @Override
                    public void onResponse(Call<Transcript> call, Response<Transcript> response) {
                        transcriptManager.set(response.body());
                        //Update the view
                        update();
                        showToolbarProgress(false);
                    }

                    @Override
                    public void onFailure(Call<Transcript> call, Throwable t) {
                        Timber.e(t, "Error refreshing the transcript");
                        showToolbarProgress(false);

                        //If this is a MinervaException, broadcast it
                        if (t instanceof MinervaException) {
                            LocalBroadcastManager.getInstance(CoursesActivity.this)
                                    .sendBroadcast(new Intent(Constants.BROADCAST_MINERVA));
                        } else {
                            DialogHelper.error(CoursesActivity.this, R.string.error_other);
                        }
                    }
                });
            }

            @Override
            public void onFailure(Call<List<Course>> call, Throwable t) {
                Timber.e(t, "Error refreshing courses");
                showToolbarProgress(false);
                //If this is a MinervaException, broadcast it
                if (t instanceof MinervaException) {
                    LocalBroadcastManager.getInstance(CoursesActivity.this)
                            .sendBroadcast(new Intent(Constants.BROADCAST_MINERVA));
                } else {
                    DialogHelper.error(CoursesActivity.this, R.string.error_other);
                }
            }
        });
    }

    /**
     * Tries to unregister from the given courses
     */
    @OnClick(R.id.course_register)
    protected void unregister() {
        //Get checked courses from adapter
        final List<Course> courses = adapter.getCheckedCourses();

        if (courses.size() > 10) {
            //Too many courses
            Utils.toast(this, R.string.courses_too_many_courses);
        } else if (courses.isEmpty()) {
            //No courses
            Utils.toast(this, R.string.courses_none_selected);
        } else {
            DialogUtils.alert(this, R.string.unregister_dialog_title,
                    R.string.unregister_dialog_message, new DialogInterface.OnClickListener() {
                        @Override
                        public void onClick(DialogInterface dialog, int which) {
                            dialog.dismiss();

                            //Don't continue if the positive button has not been clicked on
                            if (which != DialogInterface.BUTTON_POSITIVE) {
                                return;
                            }

                            //Make sure we are connected to the internet
                            if (!Utils.isConnected(CoursesActivity.this)) {
                                DialogHelper.error(CoursesActivity.this,
                                        R.string.error_no_internet);
                                return;
                            }

                            //Show the user we are loading
                            showToolbarProgress(true);

                            //Run the registration thread
                            mcGillService.registration(
                                    McGillManager.getRegistrationURL(term, courses, true))
                                    .enqueue(new Callback<List<RegistrationError>>() {
                                        @Override
                                        public void onResponse(Call<List<RegistrationError>> call,
                                                Response<List<RegistrationError>> response) {
                                            showToolbarProgress(false);

                                            //If there are no errors, show the success message
                                            if (response.body() == null ||
                                                    response.body().isEmpty()) {
                                                Utils.toast(CoursesActivity.this,
                                                        R.string.unregistration_success);
                                                return;
                                            }

                                            //Prepare the error message String
                                            String errorMessage = "";
                                            for (RegistrationError error : response.body()) {
                                                errorMessage += error.getString(courses);
                                                errorMessage += "\n";
                                            }

                                            DialogHelper.error(CoursesActivity.this, errorMessage);

                                            //Refresh the courses
                                            refresh();
                                        }

                                        @Override
                                        public void onFailure(Call<List<RegistrationError>> call,
                                                Throwable t) {
                                            Timber.e(t, "Error unregistering for courses");
                                            showToolbarProgress(false);
                                            //If this is a MinervaException, broadcast it
                                            if (t instanceof MinervaException) {
                                                LocalBroadcastManager
                                                        .getInstance(CoursesActivity.this)
                                                        .sendBroadcast(new Intent(
                                                                Constants.BROADCAST_MINERVA));
                                            } else {
                                                DialogHelper.error(CoursesActivity.this,
                                                        R.string.error_other);
                                            }
                                        }
                                    });
                        }
                    });
        }
    }

    @Override
    protected @HomepageManager.Homepage
    int getCurrentPage() {
        return HomepageManager.COURSES;
    }
}

File: app/src/main/java/ui/courses/CoursesActivity.kt
/*
 * Copyright 2014-2022 Julien Guerinet
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.guerinet.mymartlet.ui.courses

import android.os.Bundle
import android.view.Menu
import android.view.MenuItem
import android.widget.Button
import android.widget.LinearLayout
import android.widget.TextView
import androidx.core.view.isVisible
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.guerinet.mymartlet.R
import com.guerinet.mymartlet.model.Course
import com.guerinet.mymartlet.model.RegistrationError
import com.guerinet.mymartlet.model.Term
import com.guerinet.mymartlet.ui.DrawerActivity
import com.guerinet.mymartlet.ui.dialog.list.TermDialogHelper
import com.guerinet.mymartlet.util.extensions.errorDialog
import com.guerinet.mymartlet.util.manager.HomepageManager
import com.guerinet.mymartlet.util.manager.McGillManager
import com.guerinet.mymartlet.util.prefs.DefaultTermPref
import com.guerinet.mymartlet.util.retrofit.TranscriptConverter.TranscriptResponse
import com.guerinet.mymartlet.util.room.daos.CourseDao
import com.guerinet.mymartlet.util.room.daos.TranscriptDao
import com.guerinet.suitcase.coroutines.ioDispatcher
import com.guerinet.suitcase.dialog.cancelButton
import com.guerinet.suitcase.dialog.okButton
import com.guerinet.suitcase.dialog.showDialog
import com.guerinet.suitcase.util.extensions.toast
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import org.koin.android.ext.android.inject
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response

/**
 * Shows the user all of the courses they have taken or is currently registered in
 * @author Julien Guerinet
 * @author Joshua David Alfaro
 * @since 1.0.0
 */
class CoursesActivity : DrawerActivity() {

    private val defaultTermPref by inject<DefaultTermPref>()

    private val courseDao by inject<CourseDao>()

    private val transcriptDao by inject<TranscriptDao>()

    private val adapter: CoursesAdapter by lazy { CoursesAdapter(empty) }

    private var term: Term = defaultTermPref.term

    override val currentPage = HomepageManager.HomePage.COURSES

    private val empty by lazy<TextView> { findViewById(R.id.empty) }
    private val list by lazy<RecyclerView> { findViewById(R.id.list) }
    private val register by lazy<Button> { findViewById(R.id.register) }
    private val wishlist by lazy<Button> { findViewById(R.id.button) }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_wishlist)

        list.layoutManager = LinearLayoutManager(this)
        list.adapter = adapter

        // Format the unregister button
        register.setText(R.string.courses_unregister)
        register.layoutParams = LinearLayout.LayoutParams(
            LinearLayout.LayoutParams.MATCH_PARENT,
            LinearLayout.LayoutParams.WRAP_CONTENT
        )
        register.setOnClickListener { unregister() }

        // Remove the wishlist button
        wishlist.isVisible = false
    }

    override fun onResume() {
        super.onResume()
        update()
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.refresh, menu)
        menuInflater.inflate(R.menu.change_semester, menu)
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            R.id.action_change_semester -> {
                TermDialogHelper(this, this, term) {
                    // Set the default currentTerm
                    defaultTermPref.term = it

                    term = it

                    update()
                    refresh()
                }
                return true
            }
            R.id.action_refresh -> {
                refresh()
                return true
            }
            else -> return super.onOptionsItemSelected(item)
        }
    }

    /**
     * Updates all of the info in the view
     */
    private fun update() {
        launch {
            // Set the title
            title = term.getString(this@CoursesActivity)

            // User can unregister if the current currentTerm is in the list of terms to register for
            val canUnregister = withContext(ioDispatcher) {
                Term.loadRegistrationTerms().contains(term)
            }

            // Load the courses for the current term
            val courses = withContext(ioDispatcher) {
                courseDao.getTermCourses(term)
            }

            adapter.update(courses, canUnregister)

            // Change the text and the visibility if we are in the list of currently registered courses
            register.isVisible = canUnregister
        }
    }

    /**
     * Refreshes the list of courses for the given currentTerm and the user's transcript
     */
    private fun refresh() {
        if (!canRefresh()) {
            return
        }

        // Download the courses for this currentTerm
        mcGillService.schedule(term).enqueue(object : Callback<List<Course>> {
            override fun onResponse(call: Call<List<Course>>, response: Response<List<Course>>) {
                launch(Dispatchers.IO) {
                    courseDao.update(response.body() ?: listOf(), term)

                    withContext(Dispatchers.Main) {
                        // Update the view
                        update()
                        toolbarProgress.isVisible = false
                    }
                }

                // Download the transcript (if ever the user has new semesters on their transcript)
                mcGillService.oldTranscript().enqueue(object : Callback<TranscriptResponse> {
                    override fun onResponse(
                        call: Call<TranscriptResponse>,
                        response: Response<TranscriptResponse>
                    ) {
                        launch(Dispatchers.IO) {
                            transcriptDao.update(response.body()!!.transcript)
                        }
                    }

                    override fun onFailure(call: Call<TranscriptResponse>, t: Throwable) =
                        handleError("refreshing transcript", t)
                })
            }

            override fun onFailure(call: Call<List<Course>>, t: Throwable) =
                handleError("refreshing courses", t)
        })
    }

    /**
     * Attempts to unregister from the given courses
     */
    private fun unregister() {
        // Get checked courses from adapter
        val courses = adapter.checkedCourses

        if (courses.size > 10) {
            // Too many courses
            toast(getString(R.string.courses_too_many_courses))
            return
        }

        if (courses.isEmpty()) {
            // No courses
            toast(getString(R.string.courses_none_selected))
            return
        }

        showDialog(R.string.unregister_dialog_title, R.string.unregister_dialog_message) {
            cancelButton {}
            okButton {
                if (!canRefresh()) {
                    return@okButton
                }

                // Run the registration thread
                mcGillService.registration(McGillManager.getRegistrationURL(courses, true))
                    .enqueue(object : Callback<List<RegistrationError>> {
                        override fun onResponse(
                            call: Call<List<RegistrationError>>,
                            response: Response<List<RegistrationError>>
                        ) {
                            toolbarProgress.isVisible = false

                            // If there are no errors, show the success message
                            val body = response.body()
                            if (body == null || body.isEmpty()) {
                                toast(getString(R.string.unregistration_success))
                                return
                            }

                            // Prepare the error message String
                            val errorMessage = body.joinToString("\n",
                                transform = { error -> error.getString(courses) })
                            errorDialog(errorMessage)

                            // Refresh the courses
                            refresh()
                        }

                        override fun onFailure(
                            call: Call<List<RegistrationError>>,
                            t: Throwable
                        ) {
                            handleError("unregistering for courses", t)
                        }
                    })
            }
        }
    }
}
